dss/operations/lambda_params.py

```diff
@@ -0,0 +1,290 @@
+"""
+Get and set environment variables in deployed lambda functions using the SSM param store
+variable named "environment".
+"""
+import os
+import sys
+import select
+import json
+import argparse
+import logging
+import typing
+
+from botocore.exceptions import ClientError
+
+from dss.operations import dispatch
+from dss.operations.ssm_params import get_ssm_variable_prefix, fix_ssm_variable_prefix
+from dss.operations.ssm_params import get_ssm_parameter, set_ssm_parameter
+
+from dss.util.aws.clients import ssm as ssm_client  # type: ignore
+from dss.util.aws.clients import secretsmanager as sm_client  # type: ignore
+from dss.util.aws.clients import es as es_client  # type: ignore
+import dss.util.aws.clients
+lambda_client = getattr(dss.util.aws.clients, "lambda")
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_elasticsearch_endpoint() -> str:
+    domain_name = os.environ["DSS_ES_DOMAIN"]
+    domain_info = es_client.describe_elasticsearch_domain(DomainName=domain_name)
+    return domain_info["DomainStatus"]["Endpoint"]
+
+
+def get_admin_emails() -> str:
+
+    def get_secret_variable_prefix() -> str:
+        # TODO: this functionality should be moved to secrets.py
+        store_name = os.environ["DSS_SECRETS_STORE"]
+        stage_name = os.environ["DSS_DEPLOYMENT_STAGE"]
+        store_prefix = f"{store_name}/{stage_name}"
+        return store_prefix
+
+    def fix_secret_variable_prefix(secret_name: str) -> str:
+        # TODO: this functionality should be moved to secrets.py
+        prefix = get_secret_variable_prefix()
+        if not (secret_name.startswith(prefix) or secret_name.startswith("/" + prefix)):
+            secret_name = f"{prefix}/{secret_name}"
+        return secret_name
+    
+    def fetch_secret_safely(secret_name: str) -> dict:
+        # TODO: this functionality should be moved to secrets.py
+        try:
+            response = sm_client.get_secret_value(SecretId=secret_name)
+        except ClientError as e:
+            if 'Error' in e.response:
+                errtype = e.response['Error']['Code']
+                if errtype == 'ResourceNotFoundException':
```

Comment:
Should this be caught in the `try` block, i.e. `except ResourceNotFoundException as e`?

Reply:
No, because boto exceptions are weird. There are no explicitly-defined exceptions anywhere, so you cannot for example `import boto3.ResourceNotFoundException`. There is only one exception, which is the `ClientError`, and you have to pull apart the exception message to see what type of exception it was.

That being said, if you have ideas on how to make this more elegant, I would very much appreciate it.

Reply:
Related: https://gist.github.com/chmreid/1441361f70efac009a88be4def576be1

Reply:
I think you can do this by catching `sm_client.exceptions.ResourceNotFoundException` instead of `botocore.exceptions.ClientError`. Relevant issue is boto/boto3#1195. [I tried reproducing it locally](https://github.com/HumanCellAtlas/data-store/commit/b0188f848ca59a6e108cb30c31f61b7d55f20384) and it seems to work, though I had to remove references to `get_ssm_parameter` and `set_ssm_parameter` to get it to run.

Reply:
Can't get this to work, `sm_client.exceptions.ResourceNotFoundException` doesn't derive from `BaseException`. :-/